routes/index.js

```diff
@@ -150,6 +165,75 @@ module.exports = function(utils, env, nunjucksEnv, appName) {
       app.get( '/slowparse/spec/errors.forbidjs.html', function( req, res ) {
         res.render('/slowparse/spec/errors.forbidjs.html');
       });
+    },
+
+    oauth2Callback: function(req, res, next) {
+      if (req.query.logout) {
+        req.session = null;
+        return res.redirect(301, '/');
+      }
+
+      if (!req.query.code) {
+        return next({ code: 401, message: "OAUTH: Code required" });
+      }
+
+      if (!req.cookies.state || !req.query.state) {
+        return next({ code: 401, message: "OAUTH: State required" });
+      }
+
+      if (req.cookies.state !== req.query.state) {
+        return next({ code: 401, message: "OAUTH: Invalid state" });
+      }
+
+      if (req.query.client_id !== oauthClientId) {
+        return next({ code: 401, message: "OAUTH: Invalid client credentials" });
+      }
+
+      // First, fetch the token
+      request.post({
+        url: oauthUrl + '/login/oauth/access_token',
+        form: {
+          client_id: oauthClientId,
+          client_secret: oauthClientSecret,
+          grant_type: "authorization_code",
+          code: req.query.code
+        }
+      }, function(err, response, body) {
+        if (err) {
+          console.log("Request error: ", err, " Body: ", body);
+          return next({ code: 500, message: "Internal server error. See logs for details" });
+        }
+
+        if (response.statusCode !== 200) {
+          console.log("Code " + response.statusCode + ". Error getting access token: ", body);
+          return next({ code: response.statusCode, message: body });
+        }
+
+        body = JSON.parse(body);
```

Comment:
or alternatively, specify that the response type has to be JSON, and do the "this is not json" error handling in the request's error callback